File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/store/meta/RendezvousStoreGetRequest.scala
package com.webank.eggroll.core.deepspeed.store.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Deepspeed

import scala.concurrent.duration.{FiniteDuration, NANOSECONDS}
import scala.language.implicitConversions

case class RendezvousStoreGetRequest(prefix: String, key: K, timeout: FiniteDuration)

object RendezvousStoreGetRequest {
  implicit def serialize(src: RendezvousStoreGetRequest): Array[Byte] = {
    val builder = Deepspeed.StoreGetRequest.newBuilder()
      .setPrefix(src.prefix)
      .setKey(ByteString.copyFrom(src.key.toArray))
      .setTimeout(com.google.protobuf.Duration.newBuilder()
        .setSeconds(src.timeout.toSeconds)
        .setNanos(src.timeout.toNanos.toInt % 1000000000)
        .build())
    builder.build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): RendezvousStoreGetRequest = {
    val src = Deepspeed.StoreGetRequest.parseFrom(byteString)
    RendezvousStoreGetRequest(
      prefix = src.getPrefix,
      key = src.getKey.toByteArray.toVector,
      timeout = FiniteDuration(src.getTimeout.getSeconds * 1000000000L + src.getTimeout.getNanos, NANOSECONDS)
    )
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/deepspeed/store/RendezvousStoreGetRequest.java
package com.eggroll.core.deepspeed.store;

import com.eggroll.core.pojo.RpcMessage;
import com.eggroll.core.utils.JsonUtil;
import com.google.protobuf.ByteString;
import com.google.protobuf.Duration;
import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.Deepspeed;
import lombok.Data;

@Data
public class RendezvousStoreGetRequest implements RpcMessage {
    private String prefix;
    private byte[] key;
    private long timeout;

    @Override
    public byte[] serialize() {
        Deepspeed.StoreGetRequest.Builder builder = Deepspeed.StoreGetRequest.newBuilder()
                .setPrefix(this.getPrefix())
                .setKey(ByteString.copyFrom(JsonUtil.convertToByteArray(this.getKey())))
                .setTimeout(com.google.protobuf.Duration.newBuilder()
                        .setSeconds(this.getTimeout())
                        .setNanos((int) this.getTimeout())
                        .build());
        return builder.build().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Deepspeed.StoreGetRequest proto = Deepspeed.StoreGetRequest.parseFrom(data);
            this.prefix = proto.getPrefix();
            this.key = proto.getKey().toByteArray();
            this.timeout = proto.getTimeout().getSeconds();
        } catch (InvalidProtocolBufferException e) {
            e.printStackTrace();
        }
    }

}